### PR TITLE
parsers(ccpuk): add observations

### PR DESCRIPTION
**Pull request by ekamau**
First attempt at the observations table. Requires checking with the edited schema requirements.

### CODE REVIEW

File: isaric/parsers/ccpuk_observations.toml [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+{ field = "flw_survey_date", description = "About you and your COVID-19 illness (if you're completing this survey on behalf of a child or adult that you care for, all the questions relate to their health and wellbeing)" },
+{ field = "dsstdtc", description = "Outcome date", if = { dsterm = 4 } },
+{ field = "dsstdtc_v2", description = "Outcome date", if = { dsterm_v2 = 4 } }
+]
+
+[[observation]]
+name = "haemorrhage"
+date = { ref = "observationDateHierarchy" }
+phase = "admission"
+is_present = { field = "cough_ceoccur_v2", ref = "Y/N/NK" }
```
Comment (abhidg) [inline isaric/parsers/ccpuk_observations.toml:22]: something else for field?

File: isaric/parsers/ccpuk_observations.toml [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+phase = "study"
+date = { field = "daily_dsstdat", description = "Date of assessment" }
+value_num = { field = "daily_gcs_vsorres" }
+if = { daily_gcs_lbyn = 1 }
+#description = "Glasgow Coma Score (GCS/15)"
+
+[[observation]]
+name = "conjunctivitis"
+phase = "admission"
+date = { ref = admissionDateHierarchy }
```
Comment (abhidg) [inline isaric/parsers/ccpuk_observations.toml:68]: admissionDateHierarchy should be quoted

File: isaric/parsers/ccpuk_observations.toml [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+is_present = { if field = "vomit_ceoccur_v2", ref = "Y/N/NK" }
+if = { vomit_ceoccur_v2 = { "!=" = 3} }
+
+[[observation]]
+name = "vomiting"
+phase = "followup"
+date = { ref = followupDateHierarchy }
+is_present = { if field = "flw_nausea", values = { 1 = true, 0 = false } }
```
Comment (abhidg) [inline isaric/parsers/ccpuk_observations.toml:111]: if not valid here, only `{ field = "flw_nausea", ... }`

**Review by abhidg — CHANGES_REQUESTED** [review]
Some typos, otherwise looks good.

File: isaric/parsers/ccpuk_observations.toml (hunk 1 above)
Comment (ekamau) [reply isaric/parsers/ccpuk_observations.toml:22]: Yes. I think I have changed this .. or will change shortly.

File: isaric/parsers/ccpuk_observations.toml (hunk 2 above)
Comment (ekamau) [reply isaric/parsers/ccpuk_observations.toml]: I have changed this - should come through in the new PR, hopefully.

File: isaric/parsers/ccpuk_observations.toml (hunk 3 above)
Comment (ekamau) [reply isaric/parsers/ccpuk_observations.toml:123]: not sure I understand your comment/suggestion. we'll come back to this..

**Review by ekamau — COMMENTED** [review]
Thank you for the review. I added more changes in the updated files.

**Comment by pipliggins** [comment]
Is the plan to merge this file into the isaric-ccpuk.toml file once it's sorted, so we still just have a single parser file per study?

**Comment by ekamau** [comment]
YES!
That's the plan I believe, a single parser per study.

**Comment by ekamau** [comment]
We need to merge for both CCPUK and RAPID studies - all to have the three
tables in one parser, then Sadie can review ..

**Comment by pipliggins** [comment]
> YES! That's the plan I believe, a single parser per study.
> […](#)

Wow, emphatic! Cool, sounds good.

File: isaric/parsers/ccpuk_observations.toml [hunk 4]
```diff
@@ -57,424 +51,372 @@ date = { ref = "admissionDateHierarchy" }
...
+value_char = { field = "flw_eq5d_mb_p", values = { 2 = "I had slight problems in walking about", 3 = "I had moderate problems in walking about", 4 = "I had severe problems in walking about", 5 = "I was unable to walk about" } }
+if = { flw_eq5d_mb_p = { "!=" = 1 } }
 
 [[observation]]
 name = "inability_walk"
 phase = "followup"
 date = { ref = "followupDateHierarchy" }
-is_present = { if field = "flw_eq5d_mb", values = { 2 = "I had slight problems in walking about", 3 = "I had moderate problems in walking about", 4 = "I had severe problems in walking about", 5 = "I was unable to walk about" } }
+value_char = { field = "flw_eq5d_mb", values = { 2 = "I had slight problems in walking about", 3 = "I had moderate problems in walking about", 4 = "I had severe problems in walking about", 5 = "I was unable to walk about" } }
 if = { flw_eq5d_mb = { "!=" = 1 } }
```
Comment (pipliggins) [inline isaric/parsers/ccpuk_observations.toml:376]: Should one of these be for admission? duplicate otherwise

**Comment by ekamau** [comment]
Thanks for the edits.
Both mapping fields 'flw_eq5d_mb' and 'flw_eq5d_mb_p' were recorded in
followup surveys ..

**Comment by ekamau** [comment]
@abhidg  - Have a look / review this. It looks ok to me but there might be something I missed.

**Comment by abhidg** [comment]
Getting this as output, needs work on getting the date correctly

|table       	|valid	|total	|percentage_valid|
|---------------|-------|-------|----------------|
|subject       	|9221	|36754	|25.088426% |
|visit         	|30834	|36754	|83.892910% |
|observation   	|823215	|19310495	|4.263045% |

## subject

* 23190: data.age must be number (this is actually not true, bug in validation)
* 4323: data must contain ['subject_id', 'country_iso3', 'enrolment_date', 'age', 'sex_at_birth', 'ethnicity', 'pathogen'] properties
* 10: data.pregnancy_gestational_outcome must be one of ['term_birth', 'preterm_birth']
* 4: data.age must be smaller than or equal to 120
* 4: data.pregnancy_gestational_assessment_weeks must be number
* 2: data.admission_date must be date

## visit

* 5918: data must contain ['visit_id', 'subject_id', 'country_iso3', 'start_date', 'outcome', 'date_outcome'] properties
* 2: data.date_outcome must be date

## observation

* 15605818: data.date must be string (most dates are empty)
* 1712119: data.is_present must be boolean
* 1152104: data.value must be number
* 17163: data.text must be string
* 76: data.date must be date

**Comment by abhidg** [comment]
That was a bug in adtl, which will be fixed by https://github.com/globaldothealth/adtl/pull/32

After fixes:

|table       	|valid	|total	|percentage_valid|
|---------------|-------|-------|----------------|
|subject       	|32135	|36754	|87.432660% |
|visit         	|30834	|36754	|83.892910% |
|observation   	|1014060	|1025470	|98.887339% |

## subject

* 4394: data must contain ['subject_id', 'country_iso3', 'enrolment_date', 'age', 'sex_at_birth', 'ethnicity', 'pathogen'] properties
* 217: data.pregnancy_gestational_outcome must be one of ['term_birth', 'preterm_birth']
* 5: data.age must be smaller than or equal to 120
* 2: data.admission_date must be date
* 1: data.age must be bigger than 0

## visit

* 5918: data must contain ['visit_id', 'subject_id', 'country_iso3', 'start_date', 'outcome', 'date_outcome'] properties
* 2: data.date_outcome must be date

## observation

* 10362: data.date must be string
* 561: data.text must be string
* 298: data.is_present must be boolean
* 90: data.context must be array
* 61: data.date must be date
* 38: data.value must be number

visit could be improved, will merge this so that we get one almost-complete dataset ready :)

**Pull request merged**